send-transaction-service/src/send_transaction_service.rs

```diff
@@ -137,49 +169,66 @@ impl SendTransactionService {
-                            });
-                            let addresses = addresses
-                                .map(|address_list| {
-                                    if address_list.is_empty() {
-                                        vec![&tpu_address]
-                                    } else {
-                                        address_list
-                                    }
-                                })
-                                .unwrap_or_else(|| vec![&tpu_address]);
-                            for address in addresses {
-                                Self::send_transaction(address, &transaction_info.wire_transaction);
-                            }
-                            if transactions_len < MAX_TRANSACTION_QUEUE_SIZE {
-                                inc_new_counter_info!("send_transaction_service-insert-tx", 1);
-                                entry.or_insert(transaction_info);
-                            } else {
+                            if transactions_len >= MAX_TRANSACTION_QUEUE_SIZE {
+                                // queue is already full, no more queueing for retry, just send this txn out.
                                 datapoint_warn!("send_transaction_service-queue-overflow");
+                                Self::send_single_transaction(
+                                    &tpu_address,
+                                    &leader_info,
+                                    &config,
+                                    &transaction_info,
+                                );
+                            } else {
+                                entry.or_insert(transaction_info);
+                                transactions_len += 1;
+
+                                if transactions_len < config.batch_size
+                                    && last_status_check.elapsed().as_millis()
+                                        < config.batch_send_rate_ms as u128
+                                {
+                                    // Not enough transactions in the batch and not long enough time has passed since the last batch
+                                    continue;
+                                }
+                                // reached batch limit
+                                inc_new_counter_info!("send_transaction_service-batch-tx", 1);
+                                do_batch = true;
                             }
                         } else {
                             inc_new_counter_info!("send_transaction_service-recv-duplicate", 1);
                         }
                     }
                 }
 
-                if last_status_check.elapsed().as_millis() as u64 >= config.retry_rate_ms {
+                if last_status_check.elapsed().as_millis() as u64 >= config.retry_rate_ms
+                    || do_batch
+                {
                     if !transactions.is_empty() {
                         datapoint_info!(
```

Comment:
I think we need to gate submitting the TX to a fixed time rather than let the batch rate influence it

Reply:
Can you elaborate more? Note the batch rate is trying to guard against a txn being put in the batch for too long -- i.e. we want to limit the latency per txn due to this batching effect.

Reply:
The offending code doing data report is removed. That being said, I am planning to have separate PR to address the metrics to use periodical data report to replace the inc_new_counter_info.